### PR TITLE
irmin: better signature for Tree.length (port from 2.10)

### PR DESCRIPTION
**Pull request by CraigFe**
Forward port of https://github.com/mirage/irmin/pull/1676.

### CODE REVIEW

File: src/irmin/tree_intf.ml [hunk 1]
```diff
@@ -145,8 +145,15 @@ module type S = sig
   (** [find_all t k] is [Some (b, m)] if [k] is associated to the contents [b]
       and metadata [m] in [t] and [None] if [k] is not present in [t]. *)
 
-  val length : node -> int Lwt.t
-  (** [find n] is the number of entries in [n]. *)
+  val length : t -> ?cache:bool -> path -> int Lwt.t
+  (** [length t key] is the number of files and sub-nodes stored under [k] in
```
Comment (maiste) [inline src/irmin/tree_intf.ml:149]:
```suggestion
  (** [length t key] is the number of files and sub-nodes stored under [key] in
```

**Comment by maiste** [comment]
I have an issue when I compile this version of the code. Is it only on my machine?
The error comes from the `let* length = Store.Tree.length tree in` (bench/irmin-pack/trace_collection.ml l232) as the signature for `Store.Tree.length` has been modified:
```ocaml
 let create_store_after tree =
    let* watched_nodes_length =
      Lwt_list.map_s
        (fun (_, steps) ->
          let* tree = Store.Tree.find_tree tree steps in
          match tree with
          | None -> Lwt.return 0
          | Some tree -> (
              match Store.Tree.destruct tree with
              | `Node tree ->
                  let* length = Store.Tree.length tree in
                  Lwt.return length
              | `Contents _ -> Lwt.return 0))
        Def.step_list_per_watched_node
    in
    Lwt.return Def.{ watched_nodes_length }
```

File: src/irmin/tree.ml [hunk 2]
```diff
@@ -1488,7 +1488,11 @@ module Make (P : Backend.S) = struct
             | Some (`Contents _) -> Lwt.return_some `Contents
             | Some (`Node _) -> Lwt.return_some `Node))
 
-  let length = Node.length ~cache:true
+  let length t ?(cache = true) path =
+    Log.debug (fun l -> l "Tree.length %a" pp_path path);
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/tree.ml]: They should indeed, thanks.

**Comment by CraigFe** [comment]
> I have an issue when I compile this version of the code.

Thanks; not sure how I missed that one. Should be fixed now.

**Review by maiste — APPROVED** [review]
Thanks for the corrections! LGTM :+1:

**Review by Ngoguey42 — APPROVED** [review]
Thanks

**Comment by maiste** [comment]
(Because I just saw it: is it normal that you are trying to merge into `mirage:craigfe@use-new-printbox` instead of `main`?

**Comment by CraigFe** [comment]
This PR depends on [that one](https://github.com/mirage/irmin/pull/1733) for the CI to pass, so I just set the base to that to hide it from the diff while you were reviewing :-) My plan is to merge this PR once that one is merged (after setting the base back to `main`).

**Comment by CraigFe** [comment]
Thanks both for the reviews.

**Pull request merged**